core/inherents/src/lib.rs

```diff
@@ -0,0 +1,527 @@
+	pub fn put_data<I: codec::Encode>(
+		&mut self, identifier: InherentIdentifier, inherent: &I
+	) -> Result<(), RuntimeString> {
+		match self.data.insert(identifier, inherent.encode()) {
+			Some(_) => Err("Inherent with same identifier already exists!".into()),
+			None => Ok(()),
+		}
+	}
+
+	/// Replace the data for an inherent.
+	///
+	/// If it does not exist, the data is just inserted.
+	pub fn replace_data<I: codec::Encode>(
+		&mut self, identifier: InherentIdentifier, inherent: &I
+	) {
+		self.data.insert(identifier, inherent.encode());
+	}
+
+	/// Returns the data for the requested inherent.
+	///
+	/// # Return
+	///
+	/// - `Ok(Some(I))` if the data could be found and deserialized.
+	/// - `Ok(None)` if the data could not be found.
+	/// - `Err(_)` if the data could be found, but deserialization did not work.
+	pub fn get_data<I: codec::Decode>(
+		&self, identifier: &InherentIdentifier
+	) -> Result<Option<I>, RuntimeString> {
+		match self.data.get(identifier) {
+			Some(inherent) =>
+				I::decode(&mut &inherent[..])
+					.ok_or_else(|| "Could not decode requested inherent type!".into())
+					.map(Some),
+			None => Ok(None)
+		}
+	}
+}
+
+impl codec::Encode for InherentData {
+	fn encode(&self) -> Vec<u8> {
+		let keys = self.data.keys().collect::<Vec<_>>();
+		let values = self.data.values().collect::<Vec<_>>();
+
+		let mut encoded = keys.encode();
+		encoded.extend(values.encode());
+		encoded
+	}
+}
+
+impl codec::Decode for InherentData {
+	fn decode<I: codec::Input>(value: &mut I) -> Option<Self> {
+		Vec::<InherentIdentifier>::decode(value)
+			.and_then(|i| Vec::<Vec<u8>>::decode(value).map(|v| (i, v)))
+			.and_then(|(i, v)| {
+				if i.len() == v.len() {
+					Some(Self {
+						data: i.into_iter().zip(v.into_iter()).collect()
+					})
+				} else {
+					None
+				}
+			})
+	}
+}
+
+/// The result of checking inherents.
+///
+/// It either returns okay for all checks, or stores all occurred errors.
+#[derive(Encode, Decode)]
+pub struct CheckInherentsResult {
+	/// Did the check succeeded?
+	okay: bool,
+	/// Did we encountered a fatal error?
+	fatal_error: bool,
+	/// We use the `InherentData` to store our errors.
+	errors: InherentData,
+}
+
+impl CheckInherentsResult {
+	/// Create a new instance.
+	pub fn new() -> Self {
+		Self {
+			okay: true,
+			errors: InherentData::new(),
+			fatal_error: false,
+		}
+	}
+
+	/// Put an error into the result.
+	///
+	/// This makes this result resolve to `ok() == false`.
+	///
+	/// # Parameters
+	///
+	/// - identifier - The identifier of the inherent that generated the error.
+	/// - error - The error that will be encoded.
+	pub fn put_error<E: codec::Encode + IsFatalError>(
+		&mut self, identifier: InherentIdentifier, error: &E
+	) -> Result<(), RuntimeString> {
+		// Don't accept any other error
+		if self.fatal_error {
+			return Err("No other errors are accepted after an hard error!".into())
+		}
+
+		if error.is_fatal_error() {
+			// remove the other errors.
```

Comment:
Why? You can always identify that fatal error by going through the errors and ask `error.is_fatal_error()`, but having the other errors could help debugging, no?

Reply:
You could not do that, as the errors after this point are just plain `Vec<u8>`.
The reason is, that if we encounter a fatal error, checking failed a we return immediately. However, it is allowed to have multiple non-fatal errors. Something like `TimestampInFuture` is a non-fatal error.